File: Problem3_LinkedListCycle.py
# Time Complexity: O(n), where n = number of nodes in the list
# Space Complexity: O(1)
# Did this code successfully run on Leetcode : Yes

# Solution:
"""
1. Using slow(1x) and fast(2x) pointers, get the intersecting node of the pointers.
2. If there is no intersecting node, it means there is no cycle. Hence, return None.
3. If not, use 2 pointers moving at same speed (one from head and other from intersecting node) to find the start of the cycle.
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def detect_cycle(self, head: ListNode) -> ListNode:
        if not head:
           return None

        intersect_node = self.find_intersect(head)

        # If there is no intersecting node, there is no cycle. Hence, return None.
        if not intersect_node:
            return None

        # 2 pointers - one from head and other from intersecting node moving at same speed meets at the cycle start.
        p1 = head
        p2 = intersect_node

        while p1 != p2:
            p1 = p1.next
            p2 = p2.next

        return p1

    # Find the node at which slow and fast pointers intersect
    def find_intersect(self, head) -> ListNode:
        slow = fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

            if slow == fast:
                return slow

        return None

File: Problem3_LinkedListCycle.cs
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode DetectCycle(ListNode head) {
        if(head == null) return head;

        ListNode slow = head;
        ListNode fast = head.next;
        bool isCycleFound = false;

        while(fast != null && fast.next != null) {
            slow = slow.next; //Move slow pointer by 1 node
            fast = fast.next.next; //Move fast pointer by 2 nodes
            if(slow == fast) { //Until slow and fast pointers meet
                isCycleFound = true;
                fast = head; //Reset fast pointer to head
                break;
            }
        }
        if(isCycleFound) { //If there is a cycle in the list, find the node at which the cycle starts
            while(slow != fast) {
                slow = slow.next; //Move slow pointer by 1
                if(slow == fast) { //If slow and fast pointers meet, return the node
                    break;
                }
                fast = fast.next; //Else move fast pointer by 1
            }
        return slow;
        }

        return null;  //If there is no cycle found, return null
    }
}